packages/proxy/src/providers/openai.ts

```diff
@@ -66,3 +73,60 @@ export function makeFakeOpenAIStreamTransformer() {
     },
   });
 }
+
+export async function normalizeOpenAIMessages(
+  messages: ChatCompletionMessageParam[],
+): Promise<ChatCompletionMessageParam[]> {
+  return Promise.all(
+    messages.map(async (message) => {
+      console.log("message", message);
+      if (
+        message.role === "user" &&
+        message.content &&
+        typeof message.content !== "string"
+      ) {
+        message.content = await Promise.all(
+          message.content.map(
+            async (c): Promise<ChatCompletionContentPart> =>
+              await normalizeOpenAIContent(c),
+          ),
+        );
+      }
+      return message;
+    }),
+  );
+}
+
+// https://platform.openai.com/docs/guides/pdf-files?api-mode=chat
+async function normalizeOpenAIContent(
+  content: ChatCompletionContentPart,
+): Promise<ChatCompletionContentPart> {
+  if (typeof content === "string") {
+    return content;
+  }
+  switch (content.type) {
+    case "image_url":
+      console.log("image_url", content.image_url.url);
+      if (convertBase64Media(content.image_url.url)) {
+        return content;
+      } else if (content.image_url.url.endsWith(".pdf")) {
+        const base64 = await convertMediaToBase64({
+          media: content.image_url.url,
+          allowedMediaTypes: ["application/pdf"],
+          maxMediaBytes: 20 * 1024 * 1024,
+        });
+        console.log(base64.data);
+        return {
+          type: "file",
+          file: {
+            filename:
+              content.image_url.url.split("/").pop()?.split("?")[0] ??
```

Comment:
_may_ want to use `/[%\w\-. ]+\.pdf/`

i.e.
```
['http://foo.com/bar/example.pdf', 'gs://bucket/file.pdf', 'http://foo.com/bar/?file=example.pdf', 'http://foo.com/bar/#file=exa%20mple.pdf'].map(u => u.match(/[%\w\-. ]+\.pdf/)[0]).map(decodeURIComponent)
```

```
['example.pdf', 'file.pdf', 'example.pdf', 'exa mple.pdf']
```

Reply:
spruced this up + wrote some tests